### ISSUE TITLE
[BUG] Incorrect IPAddress in `config.json` for SSH access

### ISSUE THREAD

**Issue by lindhe** [rancher/rancher#43800]
**Rancher Server Setup**
- Rancher version: v2.7.6 
- Installation option (Docker install/Helm Chart): Helm chart
   - If Helm Chart, Kubernetes Cluster and version: RKE2
- Proxy/Cert Details: n/a

**Information about the Cluster**
- Kubernetes version: v1.26.8+rke2r1
- Cluster Type (Local/Downstream): downstream
   - If downstream, what type of cluster? (Custom/Imported or specify provider for Hosted/Infrastructure Provider): Infrastructure provider (VMware vSphere).

**User Information**
- What is the role of the user logged in? Admin

**Describe the bug**

In my environment, each node has two network interfaces: one to reach our internal storage network, one for external communication. The storage network has a private IP address and the external network as a public IP address.

Because the nodes are created using the VMware vSphere cluster provisioner, Rancher provides SSH access to the nodes. When I download the bundle of SSH configuration for the node by clicking "Download SSH Key", I get a ZIP file containing `config.json`, `id_rsa` and `id_rsa.pub`. The `config.json` file looks different each time I download it, specifically in the `IPAddress` field. Sometimes it will have the public IP address of the node there (which works) but sometimes it will have the private IP address of the node (which does not work for SSH). Accessing the node via SSH on the private IP is not possible; only the public IP address can be used for that purpose.

For Rancher to have a fighting change to get this right, there must be some way for us to inform Rancher about which interface is the one to use for external access (or SSH access in particular). I have not configured that, because I do not know how to. If there is a way to do so, please inform me. It might resolve this whole issue. If there is no way to configure that, I think _that_ is the root cause which should be resolved here.

**To Reproduce**

1.  Go to Cluster Management.
2. Click on the name of a cluster.
3. For a machine in a machine pool, click the menu dots and select "Download SSH Key".
4. `unzip test-9f01a06a-7phvt.zip`
5. `cat config.json`

**Result**

```json
{"IPAddress":"10.0.0.2","SSHUser":"rancher","SSHPort":22,"MachineName":"test-9f01a06a-7phvt"}
```

**Expected Result**

```json
{"IPAddress":"93.184.216.34","SSHUser":"rancher","SSHPort":22,"MachineName":"test-9f01a06a-7phvt"}
```

**Additional context**

I have also noticed that clicking the "> SSH Shell" button for a node in the Rancher GUI is unreliable. Sometimes it works, sometimes it just hangs. I bet that's caused by the same issue with the SSH config sometimes having the wrong IP address.

I've also noticed that the field `INTERNAL-IP` of the node lists the public IP address while `EXTERNAL-IP` is unpopulated. I don't know if this is related:

```shell
$ kubectl get nodes -o wide
NAME                   STATUS   ROLES                       AGE   VERSION          INTERNAL-IP      EXTERNAL-IP   OS-IMAGE                              KERNEL-VERSION                 CONTAINER-RUNTIME
test-9f01a06a-7phvt    Ready    control-plane,etcd,master   31d   v1.26.8+rke2r1   93.184.216.34   <none>        Red Hat Enterprise Linux 9.2 (Plow)   5.14.0-284.30.1.el9_2.x86_64   containerd://1.7.3-k3s1
```

**Comment by lindhe** [comment]
Maybe related to #39055

**Comment by lindhe** [comment]
Maybe related to #43610

**Comment by lindhe** [comment]
I believe I solved it! A colleague at Conoa recommended me to check the [IP address options for Rancher agent](https://ranchermanager.docs.rancher.com/v2.7/reference-guides/cluster-configuration/rancher-server-configuration/use-existing-nodes/rancher-agent-options#ip-address-options), and indeed that seems to have fixed it!

After adding the following under `.spec` for the `Cluster` resource, I am unable to reproduce my issue:

```yaml
agentEnvVars:
  - name: CATTLE_ADDRESS
     value: eth0
```

While I wish that this was documented more clearly in Rancher's documentation, it's clearly a matter of misconfiguration – not a bug in Rancher.

**Comment by lindhe** [comment]
This seems to not be fool proof. At least when I had an old cluster where `agentEnvVars` was unset, adding it later did not do the trick. But I think it seems to work for all new cluster if I set it? 🤔

**Comment by lindhe** [comment]
No. I just got the same error on a new cluster I created, where `agentEnvVars` was set as described.

I'm on a pretty old Rancher version, tough. As soon as I'm on the latest version again, I'll reopen this issue if the problem remains.

**Comment by lindhe** [comment]
I'm now on Rancher v2.10.1, yet the issue persists. 😕 Reopening.

**Comment by lindhe** [comment]
Had some issues yesterday related to this issue, so me and a colleague dug a bit deeper. We believe we have found a bug!

There's a race condition here:

https://github.com/rancher/rancher/blob/547733329c6be18d7969dd8b1c7f083636220147/package/run.sh#L41

While the script checks that the interface is up before getting the IPv4 address, it does not check that the IP address has been assigned. So if the `get_address` function is executed before the interface has been assigned an IPv4 address from the DHCP server, the `echo` statement will end up being empty instead of correctly identifying the IP address.

We believe it should be possible to configure systemd so the `rancher-system-agent` unit to depend on the IPv4 address being assigned, but we have not quite figured out the details yet.

There's also at least one more piece to this puzzle that we don't yet understand: a comment in the script says:

> If nothing is given, return empty (it will be automatically determined later if empty)

We have no idea where that logic resides, which may reveal something important about the whole ordeal.

**Comment by kwentine** [comment]
I believe this is related to, and likely resolved by:
- https://github.com/rancher/rancher/pull/54152

Turns out that when node drivers (vSphere in this case) are used for provisioning, `CATTLE_ADDRESS` and `CATTLE_INTERNAL_ADDRESS` have no effect on `node-external-ip` that is set in the RKE2 config delivered to `rancher-system-agent` on downstream nodes:

> For node-driver-based clusters, the node-external-ip and node-ip values in the configuration are derived from the node IP information that Rancher receives from rancher-machine.

After the fix, since from what I have observed the vSphere node driver only reports one IP address, `node-external-ip` will no longer be set in RKE2 configs, and only the kubelet-reported IP address will have an incidence. And hopefully that will be the reachable one!
